### DISCUSSION TITLE
Controlling via MQTT and IOBroker doesn't work

### DISCUSSION THREAD

**Discussion by bitboy00** [tbnobody/OpenDTU#989] [Q&A]
`<html>
<body>

I have here a setup with IOBroker, openDTU and a HM-600

The IOBroker gets the data from the HM-600 via MQTT and stores values in the database, which can then be displayed via Grafana. This works reliably.

Now I tried to run the communication also in the other direction. I want to be able to limit the HM-600 in power temporarily.

so I created a state under 
mqtt.1.[serialnumber].cmd.limit_nonpersistent_absolute
The topic is called
[serialnumber]/cmd/limit_nonpersistent_absolute
The type is "variable" with the content "number".

I don't use a base topic. The instance is exclusively for DTU and there I don't need it. However, I have also tried to do it with a base-topic, with the same result.

So in mqtt-explorer it looks exactly like the incoming values. 
The logfile from IOBroker shows that it also tries to subscribe to the topics.

But after that it doesn't matter which value I write into the state. The values of the DTU do not change.
I write unconfirmed and without retain. 

Is there any way to find out if DTU actually subscribes to the topic? Is there any way to see if there is an error?

Is there anything I need to be aware of that I simply overlooked?

mqtt.1 | 2023-06-02 08:40:13.189 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
-- | -- | -- | --
mqtt.1 | 2023-06-02 08:40:13.185 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
mqtt.1 | 2023-06-02 08:40:13.183 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
mqtt.1 | 2023-06-02 08:40:13.180 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
mqtt.1 | 2023-06-02 08:40:13.178 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
mqtt.1 | 2023-06-02 08:40:13.163 | info | Client [OpenDTU-11820156] publishOnSubscribe send all known states
mqtt.1 | 2023-06-02 08:40:12.980 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/restart" with regex /mqtt\.1\.[^.]*\.cmd\.restart/
mqtt.1 | 2023-06-02 08:40:12.979 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/restart" with regex /^[^\.]*\.cmd\.restart/
mqtt.1 | 2023-06-02 08:40:12.976 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/power" with regex /mqtt\.1\.[^.]*\.cmd\.power/
mqtt.1 | 2023-06-02 08:40:12.975 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/power" with regex /^[^\.]*\.cmd\.power/
mqtt.1 | 2023-06-02 08:40:12.969 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_nonpersistent_absolute" with regex /mqtt\.1\.[^.]*\.cmd\.limit_nonpersistent_absolute/
mqtt.1 | 2023-06-02 08:40:12.968 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_nonpersistent_absolute" with regex /^[^\.]*\.cmd\.limit_nonpersistent_absolute/
mqtt.1 | 2023-06-02 08:40:12.967 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_nonpersistent_relative" with regex /mqtt\.1\.[^.]*\.cmd\.limit_nonpersistent_relative/
mqtt.1 | 2023-06-02 08:40:12.966 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_nonpersistent_relative" with regex /^[^\.]*\.cmd\.limit_nonpersistent_relative/
mqtt.1 | 2023-06-02 08:40:12.964 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_persistent_absolute" with regex /mqtt\.1\.[^.]*\.cmd\.limit_persistent_absolute/
mqtt.1 | 2023-06-02 08:40:12.964 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_persistent_absolute" with regex /^[^\.]*\.cmd\.limit_persistent_absolute/
mqtt.1 | 2023-06-02 08:40:12.962 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_persistent_relative" with regex /mqtt\.1\.[^.]*\.cmd\.limit_persistent_relative/
mqtt.1 | 2023-06-02 08:40:12.960 | info | Client [OpenDTU-11820156] subscribes on "/+/cmd/limit_persistent_relative" with regex /^[^\.]*\.cmd\.limit_persistent_relative/

</body>
</html>`

**Comment by DPO99** [comment]
Please share screenshots of all mqtt settings in iobroker and dtu.

**Comment by bitboy00** [comment]
![capture_001_02062023_102430](https://github.com/tbnobody/OpenDTU/assets/102959722/5f643a58-22ff-4e4a-bf56-bd521fcc4f42)
![capture_002_02062023_102458](https://github.com/tbnobody/OpenDTU/assets/102959722/4aa85a2c-c29e-4891-9a0a-d61d74361ecb)
![capture_003_02062023_102513](https://github.com/tbnobody/OpenDTU/assets/102959722/23f36b91-444b-45ca-84e6-da903f0d6535)
![capture_004_02062023_102642](https://github.com/tbnobody/OpenDTU/assets/102959722/b7a40b0f-337f-4908-90d6-3a4530630417)

**Comment by DPO99** [comment]
Scheint das ähnliche Problem zu sein wie bei mir....ich habe es wie folgt gelöst.

-MQTT.0 = Broker
-MQTT.1 = Client

Ich steuere mit der mqtt.1 den broker mqtt.0. Das Problem ist, dass m.E. die retain einstellungen nicht ganz sauber durchlaufen und es hier zum problem mit kommt (alledings kenne ich mich hier zu wenig aus um das problem tiefer zu verstehen). ich hab das so eingestellt:

![image](https://github.com/tbnobody/OpenDTU/assets/125746530/8fd773bd-c8c2-4588-b70c-3dd050bcb3e4)

![image](https://github.com/tbnobody/OpenDTU/assets/125746530/d7de6587-62e7-4059-afcc-96defb6818b2)

Nicht schön mit den 2 instanzen aber läuft seit monaten fehlerfrei.


EDIT
siehe auch [hier](https://github.com/tbnobody/OpenDTU/discussions/874#discussioncomment-5783804)

**Reply by bitboy00** [reply]
Die Anzahl der Instanzen ist mir ziemlich egal.
Aber zur Zeit schreibe ich die gewünschten Werte einfach von Hand in den State und schon das funktioniert nicht.

**Reply by DPO99** [reply]
Schau dir das mal an #874

**Reply by bitboy00** [reply]
Den für mich erforderlichen Datenpunkt habe ich ja angelegt. 
Der wird auch geschrieben.
Im MQTT-Explorer sehe ich das auch
nur der Wechselrichter macht nichts damit.

Entweder bekommt er den nicht
oder er versteht ihn nicht
oder er hat irgendwie ein Problem mit dem Namen oder den EInstellungen des Datenpunktes.

**Reply by DPO99** [reply]
Wie gesagt ich glaube es liegt an dem Retain. Die non_persistant werte müssen ja ohne retain gesendet werden und bei mir hat es trotz setzen des hakens nicht geklappt. Irgendwo im iobroker forum hat dann einer was von diesen 2 mqtt instanzen geschrieben und das habe ich dann einfach nachgebaut.

Die Einstellungen in der MQTT Instanz sind auch merkwürdig. Es gibt eine für Mit Retain und eine für ohne Retain. Was aber passiert z.b. wenn beide haken nicht gesetzt sind? 

"Eigene States beim verbinden publizieren" musste ich auch deaktivieren, da beim Reconnect keine Steuerung der DTU möglich war.

**Reply by bitboy00** [reply]
ich verstehe nicht, was genau bei mir nicht passen soll.
Was ist das Problem bei mir mit dem Retain?

es gibt nur eine Instanz wo die DTU zugreifen kann.
Kein externer Prozess greift bisher auf diese Daten zu.
Nur die SQL-Protokollierung liest die Werte und schreibt die in die DB.

**Reply by DPO99** [reply]
Das Problem bist nicht Du, sondern vermutlich der Adapter MQTT im Iobroker und dessen Einstellungen. Ich habe aber keine andere Lösung als meine vorgeschlagene. Ggf. hat ja noch wer anders eine andere Idee.
